feat(ups_camera): add ups camera

ChrisLanks opened the pull request:
## Breaking change


## Proposed change

This change adds a new camera for UPS. When UPS delivers a package they take a picture. This will display that picture

## Type of change

<!--
  What type of change does your PR introduce?
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (which adds functionality)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Documentation update
- [ ] Adds a new shipper
- [ ] Update existing shipper

## Additional information

I also added the ability to customize the no_delivery photo for either amazon or UPS

<img width="373" height="431" alt="image" src="https://github.com/user-attachments/assets/af640bfb-23f2-4406-8958-30219df4f155" />

<img width="416" height="227" alt="image" src="https://github.com/user-attachments/assets/50b9b0cf-17d5-4ccc-acc6-c41e1a2c9bcd" />

Commit:
add ups camera

Commit:
fix: update custom workflow

Commit:
feat(ups-camera): add camera for ups

firstof9 commented:
I'll give this a good look later today.
Thanks!

firstof9 commented on tests/conftest.py:
```diff
@@ -37,8 +37,10 @@
 
 @pytest.fixture(autouse=True)
-def auto_enable_custom_integrations(enable_custom_integrations):
+def auto_enable_custom_integrations():
```
Seems this change blows up the tests. Let's add it back for now.

firstof9 replied:
```suggestion
def auto_enable_custom_integrations(enable_custom_integrations):
```

Commit:
Update tests

ChrisLanks committed:
Update tests/conftest.py

Co-authored-by: Chris <1105672+firstof9@users.noreply.github.com>

firstof9 commented:
Excellent, now it just looks like the config flow tests need to be updated with your new configurations.

Commit:
simplify tracking info for ups

ChrisLanks commented:
Ok cool. I cant get tox to run locally for some reason. I get it to run, but it errors on the home assistant custom integrations part. 

In terms of the logic, I updated it to get the UPS tracking number as well. After I get all this to pass, I do want to run this a bit. I believe tomorrow I have a UPS package being delivered so I can test in real time.

Commit:
fix timeout issue with datetime, add migration version, lint, black, fix tests

ChrisLanks commented:
OK. I had to update some tests for the amazon stuff I completed last time as I was getting some timeouts occasionally while testing.
I also bumped the version to 11, so people can use this new version without issue.
I believe this is finally complete now. 

I think real world testing is necessary with this especially with the reconfigure flow and custom image selection. I want to make sure that's buttoned up.

firstof9 commented on tests/test_helpers.py:
```diff
@@ -153,10 +153,18 @@ async def test_process_emails_external(
+
+    for remove_call in mock_osremove.call_args_list:
+        if "www/mail_and_packages/amazon/anotherfakefile.mp4" in remove_call.args[0]:
+            amazon_removed = True
```
```suggestion
            assert amazon_removed
```

firstof9 commented on tests/test_helpers.py:
```diff
@@ -153,10 +153,18 @@ async def test_process_emails_external(
+        if "www/mail_and_packages/amazon/anotherfakefile.mp4" in remove_call.args[0]:
+            amazon_removed = True
+        if "www/mail_and_packages/ups/anotherfakefile.mp4" in remove_call.args[0]:
+            ups_removed = True
```
```suggestion
            assert ups_removed
```

firstof9 commented on tests/test_helpers.py:
```diff
@@ -153,10 +153,18 @@ async def test_process_emails_external(
+        if "www/mail_and_packages/ups/anotherfakefile.mp4" in remove_call.args[0]:
+            ups_removed = True
+
+    assert amazon_removed, "Amazon file should be removed"
```
We can remove this line.

firstof9 commented on tests/test_helpers.py:
```diff
@@ -153,10 +153,18 @@ async def test_process_emails_external(
+            ups_removed = True
+
+    assert amazon_removed, "Amazon file should be removed"
+    assert ups_removed, "UPS file should be removed"
```
And this one.

ChrisLanks committed:
Update tests/test_helpers.py

Co-authored-by: Chris <1105672+firstof9@users.noreply.github.com>

ChrisLanks committed:
Update tests/test_helpers.py

Co-authored-by: Chris <1105672+firstof9@users.noreply.github.com>

Commit:
clean up lines of code

ChrisLanks commented:
I should also note that the no_deliveries for UPS still uses the amazon "no_deliveries" jpeg. There should be a custom image for UPS that says no_deliveries without AMAZON on it. That is definitely a TODO.

firstof9 commented:
Sorry I messed up that last suggestion for the test_helpers for the images. What I get for reviewing before my caffeine.

Commit:
move binary sensors to binary_sensor for circular dependency fix

Commit:
pylint

ChrisLanks commented:
No worries. I also notices that Binary_sensors should probably move to the binary_sensors.py file vs const.py to make it clearer. 

Update: added an image for UPS no deliveries, and fixed the liniting and test issues

Commit:
revert helpers_test.py logic

Commit:
add UPS no deliveries logo

Commit:
add no deliveries for ups

Commit:
update timeout for amazon packages, fix tests to new path

ChrisLanks commented:
So far, so good. My UPS package is coming tomorrow, but I will be traveling. So I'll be monitoring it tomorrow but unable to update any code.

firstof9 committed:
update sensor test

firstof9 reviewed (changes requested):
When you get time, we need tests for the new config flow options you created.
I may be able to help if I have time.

Commit:
add tests for the config flow and migration

firstof9 commented on tests/conftest.py:
```diff
@@ -1865,3 +1896,106 @@ def mock_imap_amazon_arriving_today():
         yield mock_conn
+
+
+@pytest.fixture(name="integration_v10_migration")
```
This test belongs in `test_config_flow.py`

ChrisLanks replied:
I also created a test_migrations.py. Do you think it makes sense to keep that file or would you suggest to consolidate it all to test_config_flow.py?

Commit:
correct errors for tests

Commit:
update tests for migrations and new settings

Commit:
add back original tests

Commit:
add back original tests - still broken

Commit:
add back original tests - still broken

Commit:
add back original tests - progress

ChrisLanks commented:
![Screenshot_20250819_104556_Home Assistant~2](https://github.com/user-attachments/assets/3aa208eb-f67e-4e35-8c47-013fcffac900)
![Screenshot_20250819_135954_Home Assistant~2](https://github.com/user-attachments/assets/04574edf-0d97-4ef4-bc73-9ae42708042b)

While traveling, I can confirm the ups camera is working as I expect

Commit:
update the tests to pass

ChrisLanks commented:
Finally back home. I have a package from UPS coming today. I also tested the config flow manually and it seems to work. I have found a minor bug with the strings. I am updating that now. But it appears to be working otherwise.

firstof9 commented on custom_components/mail_and_packages/config_flow.py:
```diff
@@ -739,8 +750,17 @@ async def async_step_reconfig_storage(self, user_input=None):
             self._errors, user_input = await _validate_user_input(self._data)
             if len(self._errors) == 0:
+                # Remove custom image file fields if the corresponding flags are False
+                final_data = self._data.copy()
```
Why do all this? Just pop the data out of the `self._data` dict.

ChrisLanks replied:
You are right. I had this originally because I was having issues with the tests. But then I realized it was a python version issue and I never cleaned this up.

firstof9 commented on custom_components/mail_and_packages/config_flow.py:
```diff
@@ -688,12 +697,14 @@ async def _show_reconfig_3(
-        }
+        # Defaults - only add custom image file defaults if the corresponding flags are enabled
+        defaults = {}
+        if self._data.get(CONF_CUSTOM_IMG):
```
Just add these to the `defaults` dict, no need to if them.

firstof9 commented on tests/test_config_flow.py:
```diff
@@ -10,6 +10,8 @@
 from pytest_homeassistant_custom_component.common import MockConfigEntry
 
 from custom_components.mail_and_packages.config_flow import _validate_user_input
+import custom_components.mail_and_packages.config_flow
```
Pretty sure we don't need this import.

firstof9 commented on tests/test_config_flow.py:
```diff
@@ -10,6 +10,8 @@
 
 from custom_components.mail_and_packages.config_flow import _validate_user_input
+import custom_components.mail_and_packages.config_flow
+import custom_components.mail_and_packages
```
Nor this import.

Commit:
update all languages for new custom image

Commit:
simplify the config_flow

ChrisLanks commented:
Chat GPT helped with the translations. Locally, it added amazon, ups, and modified the no mail to include USPS. The diffs look messy in github though.

ChrisLanks commented:
<img width="448" height="902" alt="image" src="https://github.com/user-attachments/assets/1431ee90-88ef-47f2-9c00-ac23a08f6a1a" />

The string change seemed to fix the reconfigure issue I was seeing. And I can see that the USPS package is still reporting correctly:

<img width="371" height="167" alt="image" src="https://github.com/user-attachments/assets/c0e53040-3ae5-4fd7-aef1-476d21cc2a83" />


Edit: missed a Spanish location for the new strings. Just added

Commit:
update last language es_419

firstof9 commented:
Excellent, let me know how your test pans out, then we can get this merged in.

ChrisLanks commented:
Im still testing. I would also appreciate anyone else who can test considering there is a good amount of line changes.
<img width="373" height="185" alt="Screenshot 2025-08-29 165928" src="https://github.com/user-attachments/assets/468dc8d6-4a19-4b1c-a682-6d736d39ea21" />

I got my USPS package and everything still seems to be ok. Its really the custom no images i'll need help testing.

ChrisLanks commented:
Update: to do some testing, i swapped the amazon and ups no package image as "custom images" and i can see in the logs its successfully identified and applied them:
```2025-08-29 20:00:07.028 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Copying /config/custom_components/mail_and_packages/mail_none.gif to /config/custom_components/mail_and_packages/images/30133fc3-cce4-46f8-94d4-f9170b9b5d08.gif
2025-08-29 20:00:07.029 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Source file exists: True
2025-08-29 20:00:07.029 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Target directory exists: True
2025-08-29 20:00:07.032 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Successfully copied image to /config/custom_components/mail_and_packages/images/30133fc3-cce4-46f8-94d4-f9170b9b5d08.gif
2025-08-29 20:00:07.032 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Target file exists after copy: True
2025-08-29 20:00:07.033 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Image name: 30133fc3-cce4-46f8-94d4-f9170b9b5d08.gif
2025-08-29 20:00:07.033 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Grid image name: 30133fc3-cce4-46f8-94d4-f9170b9b5d08_grid.png
2025-08-29 20:00:07.033 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] image_file_name called - amazon: True, ups: False
2025-08-29 20:00:07.035 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Created: 29-Aug-2025, Today: 29-Aug-2025
2025-08-29 20:00:07.037 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Image Name: 4b58b20c-d6ab-4b2c-9aff-17f90a4e940d.jpg
2025-08-29 20:00:07.037 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Copying custom_components/mail_and_packages/no_deliveries_ups.jpg to /config/custom_components/mail_and_packages/images/amazon/4b58b20c-d6ab-4b2c-9aff-17f90a4e940d.jpg
2025-08-29 20:00:07.037 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Source file exists: True
2025-08-29 20:00:07.037 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Target directory exists: True
2025-08-29 20:00:07.039 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Successfully copied image to /config/custom_components/mail_and_packages/images/amazon/4b58b20c-d6ab-4b2c-9aff-17f90a4e940d.jpg
2025-08-29 20:00:07.040 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Target file exists after copy: True
2025-08-29 20:00:07.040 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Amazon Image Name: 4b58b20c-d6ab-4b2c-9aff-17f90a4e940d.jpg
2025-08-29 20:00:07.041 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Generating UPS image name...
2025-08-29 20:00:07.041 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] image_file_name called - amazon: False, ups: True
2025-08-29 20:00:07.041 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Processing UPS image file name
2025-08-29 20:00:07.041 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Using custom UPS image: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-08-29 20:00:07.041 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] UPS path: /config/custom_components/mail_and_packages/images/ups
2025-08-29 20:00:07.044 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Created: 20-Aug-2025, Today: 29-Aug-2025
2025-08-29 20:00:07.045 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Created: 19-Aug-2025, Today: 29-Aug-2025
2025-08-29 20:00:07.046 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Created: 29-Aug-2025, Today: 29-Aug-2025
2025-08-29 20:00:07.047 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Image Name: 8bd4dae0-ef68-4e8c-83d8-55ac2c299238.jpg
2025-08-29 20:00:07.047 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Copying custom_components/mail_and_packages/no_deliveries_amazon.jpg to /config/custom_components/mail_and_packages/images/ups/8bd4dae0-ef68-4e8c-83d8-55ac2c299238.jpg
2025-08-29 20:00:07.048 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Source file exists: True
2025-08-29 20:00:07.048 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Target directory exists: True
2025-08-29 20:00:07.050 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Successfully copied image to /config/custom_components/mail_and_packages/images/ups/8bd4dae0-ef68-4e8c-83d8-55ac2c299238.jpg
2025-08-29 20:00:07.051 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Target file exists after copy: True
2025-08-29 20:00:07.051 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] UPS Image Name: 8bd4dae0-ef68-4e8c-83d8-55ac2c299238.jpg
2025-08-29 20:00:07.051 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Set ATTR_UPS_IMAGE in coordinator data: 8bd4dae0-ef68-4e8c-83d8-55ac2c299238.jpg
2025-08-29 20:00:07.051 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] UPS image file exists: /config/custom_components/mail_and_packages/images/ups/8bd4dae0-ef68-4e8c-83d8-55ac2c299238.jpg
2025-08-29 20:00:07.052 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] Image path: custom_components/mail_and_packages/images/
2025-08-29 20:00:07.052 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] if statement sensor: amazon_delivered
2025-08-29 20:00:07.052 DEBUG (SyncWorker_8) [custom_components.mail_and_packages.helpers] === PROCESSING AMAZON DELIVERED SENSOR ===
```
However, my browser didn't swap the images on my widget. Trying to determine if this is a caching issue. Have you had any luck testing @firstof9 ?

firstof9 commented:
> Trying to determine if this is a caching issue.

Very likely caching, if you hit F12 then right click the refresh button you should be able to "Empty cache and reload".

Commit:
flip if statement for custom images

Commit:
update the no_deliveries image to be proper.

ChrisLanks commented:
Ok. Got it to work. The issue was the coordinator. On Initialization, it was setting the image. When you swapped to a custom image, it would still use the coordinator material vs the custom image passed. The solution was simpler after a few attempts at trying to understand why it wasn't caching. It was to reverse the IF statements.  When there is no mail, and its a custom image, use that, otherwise, if there is no mail, use the coordinator. 
If we have a package, no mail will not be true and it will use the coordinator image. 

This is stock no custom images:
<img width="501" height="614" alt="image" src="https://github.com/user-attachments/assets/51420abe-efb8-4bad-a311-e4f4d29f34f5" />


This is after I swap the to using a custom image, and use the UPS for amazon and the amazon for UPS.  And the logs
<img width="500" height="613" alt="image" src="https://github.com/user-attachments/assets/d41ac74b-11b8-45d8-87b9-ec7eb351b114" />

```
2025-09-06 19:00:51.126 DEBUG (MainThread) [custom_components.mail_and_packages] Finished fetching Mail and Packages (imap.gmail.com) data in 11.882 seconds (success: True)
2025-09-06 19:00:51.127 DEBUG (MainThread) [custom_components.mail_and_packages.binary_sensor] binary_sensor: usps_update value: True
2025-09-06 19:00:51.127 DEBUG (MainThread) [custom_components.mail_and_packages.binary_sensor] binary_sensor: amazon_update value: True
2025-09-06 19:00:51.128 DEBUG (MainThread) [custom_components.mail_and_packages.camera] UPS camera - custom image enabled: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:00:51.128 DEBUG (MainThread) [custom_components.mail_and_packages.camera] UPS camera - initial file path set to: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:00:51.128 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Amazon camera - custom image enabled: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:00:51.128 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Amazon camera - initial file path set to: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:00:51.138 DEBUG (MainThread) [custom_components.mail_and_packages.sensor] Updating system image path to: custom_components/mail_and_packages/images/
2025-09-06 19:00:51.140 DEBUG (MainThread) [custom_components.mail_and_packages.config_flow] mail_and_packages reconfigured.
2025-09-06 19:00:53.447 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/mail_none.gif
2025-09-06 19:01:03.440 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/mail_none.gif
2025-09-06 19:01:13.444 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/mail_none.gif
2025-09-06 19:01:21.130 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Update: usps_camera
2025-09-06 19:01:21.130 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Custom No Mail: None
2025-09-06 19:01:21.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Update: ups_camera
2025-09-06 19:01:21.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Custom No Mail: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:01:21.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] UPS camera - using custom no mail: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:01:21.133 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Update: amazon_camera
2025-09-06 19:01:21.133 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Custom No Mail: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:01:21.133 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Amazon camera - using custom no mail image: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:01:23.461 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/images/344c6352-9ba8-4d7a-b1bd-332ac17e81e0.gif
2025-09-06 19:01:33.450 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/images/344c6352-9ba8-4d7a-b1bd-332ac17e81e0.gif
2025-09-06 19:01:41.617 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail Amazon Delivery Camera attempting to read image from: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:01:41.618 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/images/344c6352-9ba8-4d7a-b1bd-332ac17e81e0.gif
2025-09-06 19:01:41.625 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail UPS Camera attempting to read image from: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:01:43.460 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/images/344c6352-9ba8-4d7a-b1bd-332ac17e81e0.gif
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Update: usps_camera
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Custom No Mail: None
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Update: ups_camera
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Custom No Mail: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] UPS camera - using custom no mail: custom_components/mail_and_packages/no_deliveries_amazon.jpg
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Update: amazon_camera
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Custom No Mail: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:01:51.131 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Amazon camera - using custom no mail image: custom_components/mail_and_packages/no_deliveries_ups.jpg
2025-09-06 19:01:53.454 DEBUG (MainThread) [custom_components.mail_and_packages.camera] Camera Mail USPS Camera attempting to read image from: /config/custom_components/mail_and_packages/images/344c6352-9ba8-4d7a-b1bd-332ac17e81e0.gif
```

Now its working!

firstof9 commented:
Excellent! I would have tested myself but I'm very time constrained anymore.

ChrisLanks commented:
Same. But we got there! I'll leave the ball in your court. We can merge and refine if necessary, or we can let it bake for a week. I have no new packages except 1 wednesday. Not sure if its coming as UPS or USPS though.
